### PR TITLE
Add metaschema models for constraints

### PR DESCRIPTION
**Pull request by david-waltermire**
# Committer Notes

Add and integrate Metaschema-based models for the legacy and [new](https://github.com/usnistgov/metaschema/blob/develop/schema/xml/metaschema-meta-constraints.xsd) Metaschema external constraints formats.

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

### CODE REVIEW

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java [hunk 1]
```diff
@@ -249,18 +250,44 @@ private IAssemblyNodeItem getCycledInstance(
    * @return the stream of descendant instances
    */
   @NonNull
-  protected Stream<INamedModelInstanceAbsolute> getNamedModelInstances(@NonNull IContainerModelAbsolute container) {
+  protected Stream<? extends INamedModelInstance> getNamedModelInstances(@NonNull IContainerModelAbsolute container) {
```
Comment (aj-stein-nist) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java:253]: Non-blocking: sorry to ask conceptual questions and we can take this to Element/Gitter if you'd prefer, but without more code context for me to understand I am struggling to understand: what makes a container of field "absolute" here?

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/xml/XmlMetaConstraintLoader.java [hunk 2]
```diff
@@ -61,10 +62,11 @@
...
 import edu.umd.cs.findbugs.annotations.Nullable;
 
-public class MetaConstraintLoader
-    extends AbstractLoader<IConstraintSet> {
+public class XmlMetaConstraintLoader
+    extends AbstractLoader<IConstraintSet>
+    implements IConstraintLoader {
   // private static final Logger LOGGER =
-  // LogManager.getLogger(MetaConstraintLoader.class);
+  // LogManager.getLogger(XmlMetaConstraintLoader.class);
```
Comment (aj-stein-nist) [inline core/src/main/java/gov/nist/secauto/metaschema/core/model/xml/XmlMetaConstraintLoader.java]: Non-blocking: is there a reason we don't want to implement the logger pattern on this class?

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java [hunk 3]
```diff
@@ -40,7 +38,7 @@
  *          the Java type of the constraint container
  */
 public abstract class AbstractTargetedConstraints<T extends IValueConstrained>
```
Comment (aj-stein-nist) [inline core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java:40]: Non-blocking: I will run through the code and try to make sure I understand better while I do a more thorough review, but is there such a thing with external constraints an "untargeted" constraint. Reading the Javadocs in another class:

```java
 * Represents a set of constraints that target a given definition using a target
 * Metapath expression.
```

Is there an alternative or opposite to this situation such we need to name this? I think I am missing something obvious but wanted to check.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java (hunk 1 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java:253]: I have been strugling with the naming here. There are two types of `INamedModelInstance`: `INamedModelInstanceAbsolute` and `INamedModelInstanceGrouped`. The latter are modeled as part of a choice group, while the former are directly part of the model. An `INamedModelInstance` represents both.

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/xml/XmlMetaConstraintLoader.java (hunk 2 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/model/xml/XmlMetaConstraintLoader.java]: There is nothing that is logged on this class. These lines can be removed.

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java (hunk 3 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java:40]: All constraints are targeted at a definition. Do you have a suggestion for a better name?

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java (hunk 1 above)
Comment (aj-stein-nist) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java:253]: OK thank you just wanted to make sure I got it. If I can think of tactical documentation recommendations to help present-tense A.J. and future A.J.'s (plural), I will mull it over for tomorrow morning.

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java (hunk 3 above)
Comment (aj-stein-nist) [reply core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java:40]: Let me mull this over too tonight, but probably not. I guess it is "a given" but maybe that is a good reason to keep it in the name. Maybe there is some tactical addition to class documentation to help that, and I will look at the Javadocs and think about if adding something to this base class will help me and those like me in the future. If not, I will resolve and close.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java (hunk 1 above)
Comment (aj-stein-nist) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/node/AbstractNodeItemFactory.java:253]: If I want document this the classes are diffed and part of the PR for the scope of review, so I will have to revisit this in a separate PR if I want to do this. I'll suppress it for now.

File: core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java (hunk 3 above)
Comment (aj-stein-nist) [reply core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/AbstractTargetedConstraints.java:40]: Ditto here, I guess I am good for now.

**Review by aj-stein-nist — APPROVED** [review]

**Pull request merged**